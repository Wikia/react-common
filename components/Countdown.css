@keyframes countdown-animation {
  from {
    stroke-dashoffset: 0;
  }
  to {
    stroke-dashoffset: 95px;
  }
}
.countdown__wrapper {
  align-items: center;
  background-color: #fff;
  border-radius: 100%;
  color: #1a1a1a;
  display: inline-flex;
  height: 40px;
  justify-content: center;
  position: relative;
  width: 40px;
}
.countdown__wrapper::after {
  border: 1px solid #7f9998;
  border-radius: 100%;
  content: " ";
  height: 30px;
  left: 4px;
  position: absolute;
  top: 4px;
  width: 30px;
}
.countdown__value {
  color: #002a32;
  font-size: 18px;
  font-weight: 700;
}
.countdown__animation {
  height: 34px;
  position: absolute;
  right: 3px;
  top: 3px;
  transform: rotateY(-180deg) rotateZ(-90deg);
  width: 34px;
  z-index: 1;
}
.countdown__animation.is-low {
  stroke: #e81a3f;
}
.countdown__animation.is-medium {
  stroke: #de5e33;
}
.countdown__animation.is-high {
  stroke: #018c30;
}
.countdown__animation > circle {
  fill: none;
  stroke-dasharray: 95px;
  stroke-dashoffset: 0;
  stroke-width: 3px;
}
.countdown__animation > circle:not(.is-stopped) {
  animation: countdown-animation 10s linear forwards;
}