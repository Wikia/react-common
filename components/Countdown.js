'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) {
    return;
  }

  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance");
}

/**
 * Why useInterval hook?
 * @see https://overreacted.io/making-setinterval-declarative-with-react-hooks/
 */

var useInterval = function useInterval(callback, delay) {
  var savedCallback = React.useRef(); // Remember the latest callback.

  React.useEffect(function () {
    savedCallback.current = callback;
  }, [callback]); // Set up the interval.

  React.useEffect(function () {
    if (delay !== null) {
      var id = setInterval(function () {
        return savedCallback.current();
      }, delay);
      return function () {
        return clearInterval(id);
      };
    }

    return undefined;
  }, [delay]);
};

var ONE_SECOND = 1000;
var START = 10;
var HIGH = 7;
var LOW = 3;
/**
 * Simple circular, countdown-from-10 component with callback.
 */

var Countdown = function Countdown(_ref) {
  var className = _ref.className,
      stop = _ref.stop,
      onTick = _ref.onTick;

  var _useState = React.useState(START),
      _useState2 = _slicedToArray(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  var timerShouldPlay = !stop && value > 0;
  useInterval(function () {
    var newValue = value - 1;
    setValue(newValue);
    onTick(newValue);
  }, timerShouldPlay ? ONE_SECOND : null); // eslint-disable-next-line

  var rangeClassName = value <= LOW ? 'is-low' : value <= HIGH ? 'is-medium' : 'is-high';
  return React__default.createElement("div", {
    className: "countdown__wrapper ".concat(className)
  }, React__default.createElement("span", {
    className: "countdown__value"
  }, value), value > 0 && React__default.createElement("svg", {
    className: "countdown__animation ".concat(rangeClassName)
  }, React__default.createElement("circle", {
    className: timerShouldPlay ? '' : 'is-stopped',
    r: "15",
    cx: "17",
    cy: "17"
  })));
};

Countdown.propTypes = {
  /** Additional class name */
  className: PropTypes.string,

  /** Callback function that will be called at every tick - 1st param is the value */
  onTick: PropTypes.func.isRequired,

  /** if `true` the countdown will be stopped */
  stop: PropTypes.bool
};
Countdown.defaultProps = {
  className: '',
  stop: false
};

module.exports = Countdown;
