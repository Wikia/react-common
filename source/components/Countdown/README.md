Click the theme buttons to re-initialize examples.

```js
<Countdown onTick={(value) => console.log('Countdown: ', value)} />
```
