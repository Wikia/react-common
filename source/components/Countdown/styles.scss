@import "~design-system/dist/scss/wds-variables/index";

@keyframes countdown-animation {
	from {
		stroke-dashoffset: 0;
	}

	to {
		stroke-dashoffset: 95px;
	}
}

.countdown {
	&__wrapper {
		align-items: center;
		background-color: $wds-color-white;
		border-radius: 100%;
		color: $wds-color-black;
		display: inline-flex;
		height: 40px;
		justify-content: center;
		position: relative;
		width: 40px;

		&::after {
			border: 1px solid $wds-fandom-color-middle-gray;
			border-radius: 100%;
			content: ' ';
			height: 30px;
			left: 4px;
			position: absolute;
			top: 4px;
			width: 30px;
		}
	}

	&__value {
		color: $wds-fandom-color-navy;
		font-size: 18px;
		font-weight: $wds-font-weight-bold;
	}

	&__animation {
		height: 34px;
		position: absolute;
		right: 3px;
		top: 3px;
		transform: rotateY(-180deg) rotateZ(-90deg);
		width: 34px;
		z-index: 1;

		&.is-low {
			stroke: $wds-color-alert;
		}

		&.is-medium {
			stroke: $wds-color-warning;
		}

		&.is-high {
			stroke: $wds-color-success;
		}
	}

	&__animation > circle {
		animation: countdown-animation 10s linear forwards;
		fill: none;
		stroke-dasharray: 95px;
		stroke-dashoffset: 0;
		stroke-width: 3px;
	}
}
