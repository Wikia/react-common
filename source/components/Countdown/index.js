import React, { useState } from 'react';
import PropTypes from 'prop-types';

import useInterval from '../../hooks/useInterval';

import './styles.scss';

const ONE_SECOND = 1000;
const START = 10;
const HIGH = 7;
const LOW = 3;

/**
 * Simple circular, countdown-from-10 component with callback.
 */
const Countdown = ({ onTick }) => {
    const [value, setValue] = useState(START);

    useInterval(() => {
        const newValue = value - 1;
        setValue(newValue);
        onTick(newValue);
    }, value > 0 ? ONE_SECOND : null);

    // eslint-disable-next-line
    const rangeClassName = (value < LOW ? 'is-low' : (value < HIGH ? 'is-medium' : 'is-high'));

    return (
        <div className="countdown__wrapper">
            <span className="countdown__value">{value}</span>
            {value > 0 && (
                <svg className={`countdown__animation ${rangeClassName}`}>
                    <circle r="15" cx="17" cy="17" />
                </svg>
            )}
        </div>
    );
};

Countdown.propTypes = {
    /** Callback function that will be called at every tick - 1st param is the value */
    onTick: PropTypes.func.isRequired,
};

export default Countdown;
