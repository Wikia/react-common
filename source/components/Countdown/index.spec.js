import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import sinon from 'sinon';

import Countdown from './index';

jest.useFakeTimers();

test('Countdown renders correctly', () => {
    const callback = sinon.spy();
    const component = mount(<Countdown onTick={callback} />);
    expect(component).toMatchSnapshot();
});

describe('Countdown calls callback 10 times with proper values', () => {
    const callback = sinon.spy();
    const component = mount(<Countdown onTick={callback} />);

    for (let value = 9; value >= 0; value -= 1) {
        const index = 9 - value;
        test(`Callback #${index}`, () => {
            act(() => {
                jest.advanceTimersByTime(1000);
            });
            component.update();

            expect(callback.callCount).toEqual(index + 1);
            expect(callback.getCall(index).args[0]).toEqual(value);
            expect(component).toMatchSnapshot();
        });
    }
});
