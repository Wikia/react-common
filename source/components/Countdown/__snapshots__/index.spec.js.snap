// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown calls callback 10 times with proper values Callback #0 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      9
    </span>
    <svg
      className="countdown__animation is-high"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #1 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      8
    </span>
    <svg
      className="countdown__animation is-high"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #2 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      7
    </span>
    <svg
      className="countdown__animation is-medium"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #3 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      6
    </span>
    <svg
      className="countdown__animation is-medium"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #4 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      5
    </span>
    <svg
      className="countdown__animation is-medium"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #5 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      4
    </span>
    <svg
      className="countdown__animation is-medium"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #6 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      3
    </span>
    <svg
      className="countdown__animation is-low"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #7 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      2
    </span>
    <svg
      className="countdown__animation is-low"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #8 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      1
    </span>
    <svg
      className="countdown__animation is-low"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;

exports[`Countdown calls callback 10 times with proper values Callback #9 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      0
    </span>
  </div>
</Countdown>
`;

exports[`Countdown renders correctly 1`] = `
<Countdown
  className=""
  onTick={[Function]}
>
  <div
    className="countdown__wrapper "
  >
    <span
      className="countdown__value"
    >
      10
    </span>
    <svg
      className="countdown__animation is-high"
    >
      <circle
        cx="17"
        cy="17"
        r="15"
      />
    </svg>
  </div>
</Countdown>
`;
